### PR TITLE
Port 80X EGM regression in 10X

### PR DESCRIPTION
**Pull request by jainshilpi**
This is just the same PR as  #22410 except that this request is for the master branch (#22410 was for 94X) .

This pull request ports the 80X EGM regression in 10X. It was reported by Htogg people that they faced problems in using 80X regression in 94X (i.e. loading an object that reloads database info and reevaluates information). This was because the EGM regression changed in 9X. So EGMPOG now plans to maintain the previous versions of regression as well. Changes done:
(1) Renamed : RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc → RecoEgamma/EgammaTools/plugins/EGRegressionModifierV2.cc
(This contains the latest version of corrections)

(2) Make a new class RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc → This is essentially the class which has 80X regression.

(3) RecoEgamma/EgammaTools/python/regressionModifier_cfi.py → This now contains a cms.PSet named regressionModifier80X which contains all the parameters set for 80X regression.

This setup has been tested on 80X root file in 10X and the reloading of database info and re-evaluating information is running fine now.

Thanks,
Shilpi

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
@cmsbuild please test

to refresh the tests

**Comment by slava77** [comment]
this could be a good chance to brush up the style of this and the V2 modifiers
I'll make comments and we see how much can be done quickly

File: RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc [hunk 1]
```diff
@@ -0,0 +1,667 @@
...
+  
+  // just calls reco versions
+  void modifyObject(pat::Electron&) const final; 
+  void modifyObject(pat::Photon&) const final;
+
+private:
+  electron_config e_conf;
+  photon_config   ph_conf;
+  std::unordered_map<unsigned,edm::Ptr<reco::GsfElectron> > eles_by_oop; // indexed by original object ptr
+  std::unordered_map<unsigned,edm::Handle<edm::ValueMap<float> > > ele_vmaps;
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc]: please switch to the same style of local variables to use trailing underscore in the name

File: RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc [hunk 2]
```diff
@@ -0,0 +1,667 @@
...
+    for( const std::string& name : parameters ) {
+      if( std::string(electronSrc) == name ) 
+	continue;
+      if( electrons.existsAs<edm::InputTag>(name)) {
+	for (auto vmp : intValueMaps) {
+	  if (name == vmp) {
+	    e_conf.tag_int_token_map[name] = ValMapIntTagTokenPair(electrons.getParameter<edm::InputTag>(name), ValMapIntToken());
+	    break;
+	  } 
+	}
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc]: this loop looks simply as 
```c++
if (intValueMaps.find(name) != intValueMaps.end()) {
e_conf.tag_int_token_map[name] = {electrons.getParameter<edm::InputTag>(name), ValMapIntToken()};
}
```

File: RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc [hunk 3]
```diff
@@ -0,0 +1,667 @@
...
+  float clusterMaxDRDPhi = 999.;
+  float clusterMaxDRDEta = 999.;
+  float clusterMaxDRRawEnergy = 0.;
+  
+  size_t iclus = 0;
+  float maxDR = 0;
+  edm::Ptr<reco::CaloCluster> pclus;
+  // loop over all clusters that aren't the seed  
+  auto clusend = the_sc->clustersEnd();
+  for( auto clus = the_sc->clustersBegin(); clus != clusend; ++clus ) {
```
[the remark this answers is not in the document]
Comment (slava77) [reply RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc]: actually, better yet, drop ```edm::Ptr<reco::CaloCluster> pclus; ``` outside of the loop and simply use here
```for (auto const& pclus : the_sc->clusters())```

File: RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc [hunk 4]
```diff
@@ -0,0 +1,667 @@
...
+  else 
+    _ecalLocal.localCoordsEE(*theseed, *iSetup_, cryEta, cryPhi, iEta, iPhi, dummy, dummy);
+
+  if (iseb) {
+    eval[29] = cryEta;
+    eval[30] = cryPhi;
+    eval[31] = iEta;
+    eval[32] = iPhi;
+  } else {
+    eval[29] = the_sc->preshowerEnergy()/the_sc->rawEnergy();
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc]: ```/raw_energy``` to use already computed const

File: RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc [hunk 5]
```diff
@@ -0,0 +1,667 @@
...
+  if (!iseb)  
+    ecor = mean*(eval[1]+the_sc->preshowerEnergy());
+  const double sigmacor = sigma*ecor;
+  
+  ele.setCorrectedEcalEnergy(ecor);
+  ele.setCorrectedEcalEnergyError(sigmacor);
+    
+  // E-p combination 
+  //std::array<float, 11> eval_ep;
+  float eval_ep[11];
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc]: actually switch to std::array (as already done in V2 modifier)

File: RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc [hunk 6]
```diff
@@ -0,0 +1,667 @@
...
+      weight = 1.;
+    else if(weight<0.) 
+      weight = 0.;
+  }
+
+  double combinedMomentum = weight*ele.trackMomentumAtVtx().R() + (1.-weight)*ecor;
+  double combinedMomentumError = sqrt(weight*weight*ele.trackMomentumError()*ele.trackMomentumError() + (1.-weight)*(1.-weight)*sigmacor*sigmacor);
+
+  math::XYZTLorentzVector oldMomentum = ele.p4();
+  math::XYZTLorentzVector newMomentum = math::XYZTLorentzVector(oldMomentum.x()*combinedMomentum/oldMomentum.t(),
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc]: compute ```corr = combinedMomentum/oldMomentum.t()``` outside and reuse it.

File: RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc [hunk 7]
```diff
@@ -0,0 +1,667 @@
...
+								oldMomentum.y()*combinedMomentum/oldMomentum.t(),
+								oldMomentum.z()*combinedMomentum/oldMomentum.t(),
+								combinedMomentum);
+ 
+  //ele.correctEcalEnergy(combinedMomentum, combinedMomentumError);
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc]: this comment can be removed by now

File: RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc [hunk 8]
```diff
@@ -0,0 +1,667 @@
...
+    EBDetId ebseedid(theseed->seed());
+    eval[26] = pho.e5x5()/theseed->energy();
+    int ieta = ebseedid.ieta();
+    int iphi = ebseedid.iphi();
+    eval[27] = ieta;
+    eval[28] = iphi;
+    int signieta = ieta > 0 ? +1 : -1; /// this is 1*abs(ieta)/ieta in original training
+    eval[29] = (ieta-signieta)%5;
+    eval[30] = (iphi-1)%2;
+    //    eval[31] = (abs(ieta)<=25)*((ieta-signieta)%25) + (abs(ieta)>25)*((ieta-26*signieta)%20); //%25 is unnescessary in this formula
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGRegressionModifierV1.cc]: this comment is likely irrelevant now

**Review by slava77 — COMMENTED** [review]
changes suggested for V1 in a number of cases also apply to V2.
Please check.

**Comment by slava77** [comment]
@jainshilpi 
are you available to follow up on the code review soon?
I have mostly used this PR as an opportunity to record where parts of the code can be improved.

If these modifiers are a dead-end in development (from the discussion in the scale/smearing PR), it's fine to leave for later. The next "V3" or even V2 modifications will likely reopen this request though.

**Comment by jainshilpi** [comment]
Hi @slava77 

I would like to take this opportunity and help with the code improvement. But I will be able to get back to it tomorrow. sorry for this delay. 

Thank you,

**Comment by slava77** [comment]
> I would like to take this opportunity and help with the code improvement. But I will be able to get back to it tomorrow. sorry for this delay.

based on the exchange on skype, this is not going to happen now and this code may even undergo even more significant changes.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/22417	112c644
- changes are in line with the PR description; this is intended to not affect the standard workflows and to bring back the corrector/modifier used in 80X
- jenkins tests pass and comparisons with the baseline show no differences

#22655 is created to follow up on the code

**Pull request merged**